apps/rush-lib/src/logic/taskRunner/AsyncTaskQueue.ts

```diff
@@ -0,0 +1,160 @@
+// Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license.
+// See LICENSE in the project root for license information.
+
+import { Task } from './Task';
+import { TaskStatus } from './TaskStatus';
+
+/**
+ * @internal
+ *
+ * Implmentation of the async iteration protocol for a collection of Task objects.
+ * The async iterator will wait for a task to be ready for execution, or terminate if there are no more tasks.
+ *
+ * @remarks
+ * It is imperative that the caller update the status of tasks prior to invoking `next()` on the iterator again,
+ * otherwise it will stall forever unless some other call stack invokes next().
+ */
+export class AsyncTaskQueue implements AsyncIterable<Task>, AsyncIterator<Task> {
+  private readonly _queue: Task[];
+  private readonly _pendingIterators: ((result: IteratorResult<Task>) => void)[];
+
+  /**
+   * @param tasks - The set of tasks to be executed
+   * @param taskSortFn - A function that sorts tasks in reverse priority order:
+   *   - Returning a positive value indicates that `a` should execute before `b`.
+   *   - Returning a negative value indicates that `b` should execute before `a`.
+   *   - Returning 0 indicates no preference.
+   */
+  public constructor(tasks: Iterable<Task>, taskSortFn: ITaskSortFunction) {
+    this._queue = computeTopologyAndSort(tasks, taskSortFn);
+    this._pendingIterators = [];
+  }
+
+  /**
+   * For use with `for await (const task of taskQueue)`
+   * @see {AsyncIterator}
+   */
+  public next(): Promise<IteratorResult<Task>> {
+    const { _queue: taskQueue, _pendingIterators: waitingIterators } = this;
+
+    const promise: Promise<IteratorResult<Task>> = new Promise(
+      (resolve: (result: IteratorResult<Task>) => void) => {
+        waitingIterators.push(resolve);
+      }
+    );
+
+    // By iterating in reverse order we do less array shuffling when removing tasks
+    for (let i: number = taskQueue.length - 1; waitingIterators.length > 0 && i >= 0; i--) {
+      const task: Task = taskQueue[i];
+
+      if (task.status !== TaskStatus.Ready) {
+        // It shouldn't be on the queue, remove it
+        // This should be a blocked task
```

Comment:
That would make sense. Technically tasks in the queue should only exist in the states "Ready" and "Blocked", because the former is the initial state and the latter is the only one that gets set by execution of other tasks. However, this particular check was migrated directly from the old code. I can update it, but it's worth checking that tasks only are in one of those two states.

Reply:
I'd say it's fine as-is then, but yes checking for only those two states would be a good call. Otherwise it should blow up with an explicit error message. Would make it clearer when someone tries to re-call `next()` what exactly happened.